Added a model to store who overrode a score

davestgermain opened the pull request:
[EDUCATOR-4295](https://openedx.atlassian.net/browse/EDUCATOR-4295)

matthugs commented on lms/djangoapps/grades/models.py:
```diff
@@ -791,3 +791,15 @@ def __unicode__(self):
+
+class ScoreOverrider(models.Model):
+    """
+    Records who overrode a score, for bulk score assignments.
```
Could you expand this with a little architectural decision record-y language around why not add this as a column on the corresponding table? e.g. just "we took this approach b/c adding columns to cwsm would be cray"

davestgermain replied:
How about [an ADR](https://github.com/edx/edx-platform/blob/8620b50396a29e093de2ba6730ce8bac9b4bc997/lms/djangoapps/grades/docs/decisions/0007-score-override-audit.rst)?

iloveagent57 replied:
A philosophical question: is the `grades` djangoapp/plugin the right place for this?  It doesn't relate to any of the other grade models, and most of the grading functions don't depend on it at all - only the csv processing stuff in grade_utils.py uses it.

davestgermain commented on lms/djangoapps/grades/grade_utils.py:
```diff
@@ -15,8 +15,9 @@
 from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
 from student.models import CourseEnrollment
-from super_csv import ChecksumMixin, CSVProcessor, DeferrableMixin
+from super_csv.csv_processor import ChecksumMixin, CSVProcessor, DeferrableMixin
```
It works either way, but this should have a version bump. Thanks for catching that. 
Also I need to move that repo to the edx org.

iloveagent57 commented on lms/djangoapps/grades/grade_utils.py:
```diff
@@ -211,4 +217,7 @@ def get_scores(usage_key, user_ids=None):
         }
+        last_override = row.scoreoverrider_set.latest('created')
+        if last_override:
+            scores[row.student_id]['who_last_graded'] = last_override.user_id
```
Should this actually be `user.id`, since the ScoreOverride models has a `user` field?

davestgermain replied:
`user.id` will do a query for the auth_user. In this case, none of the other information on the user is necessary.

iloveagent57 commented on lms/djangoapps/grades/grade_utils.py:
```diff
@@ -159,11 +161,15 @@ def set_score(usage_key, student_id, score, max_points, **defaults):
+    if override_user_id:
+        ScoreOverrider.objects.create(
+            module=module,
+            user_id=override_user_id)
```
Nit: this closing paren should be on its own line.

iloveagent57 commented on lms/djangoapps/grades/models.py:
```diff
@@ -791,3 +791,18 @@ def __unicode__(self):
+    """
+    module = models.ForeignKey(StudentModule, on_delete=models.CASCADE)
+    created = models.DateTimeField(auto_now_add=True, db_index=True)
+    user = models.ForeignKey(User, blank=True, null=True, on_delete=models.CASCADE)
```
Is it useful to store a row about a score change if the changing user is null?

nasthagiri commented on lms/djangoapps/grades/grade_utils.py:
```diff
@@ -15,8 +15,9 @@
 from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
 from student.models import CourseEnrollment
-from super_csv import ChecksumMixin, CSVProcessor, DeferrableMixin
+from super_csv.csv_processor import ChecksumMixin, CSVProcessor, DeferrableMixin
```
Hi there.  Folks from the BOM squad now automatically get notified when changes happen in certain parts of our core platform, per [CODEOWNERS](https://github.com/edx/edx-platform/blob/master/.github/CODEOWNERS).

May I kindly request that we consider keeping the core Grades layer as lean as possible so that it only owns the business logic for computing, storing, and retrieving grades.  This is the intention of the grades django app, as now described in its [README](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/README.rst).

To that end, features that export and format Grades in CSVs should be implemented in a separate layer/app.

davestgermain commented:
@nasthagiri @iloveagent57 I refactored this bulk scoring/grading into a new django app -- https://github.com/davestgermain/bulk-grades

I understand the desire to stop adding stuff to edx-platform, but it's difficult to add apps which depend on code in edx-platform. Is there any work being done to improve this process? In `bulk-grades` I added these minimal [mock django apps](https://github.com/davestgermain/bulk-grades/tree/master/mock_apps) from edx-platform in order for tests to work (I haven't written the tests yet).
